### ISSUE TITLE
Dependencies of dependencies not found while ldd does find them

### ISSUE THREAD

**Issue by haampie** [linuxdeploy/linuxdeploy#76]
`linuxdeploy` should work iff `ldd -d executable` manages to (recursively) locate all dependencies, right? I found an example where it does not.

I've upgraded my development environment from Ubuntu 16 -> 18 and thereby also upgraded the default compiler gcc 5 -> 7. Everything's fine, except that apparently gcc has changed the behaviour of `-rpath` somewhat: basically where I previously had RPATH in my binaries I now have RUNPATH.

This is fine, since RPATH was deprecated long ago, but the difference is that RUNPATH is not inherited, and this leads to an issue with `linuxdeploy`.

My binary looks like this:

```
$ readelf -d example | grep path
 0x000000000000001d (RUNPATH)            Library runpath: [...:/home/user/Qt/5.12.3/gcc_64/lib]
```

`ldd` locates all dependencies just fine:

```
$ ldd -d example | grep "not found"
# empty
```

and I can run the application from the terminal just fine without specifying `LD_LIBRARY_PATH`.

However, `linuxdeploy` fails:

```
$ QML_SOURCES_PATHS=/path/to/resources/qml ./linuxdeploy-x86_64.AppImage -v0 --appdir Example.AppDir -e Example.AppDir/usr/bin/example -i Example.AppDir/example.png -d Example.AppDir/example.desktop --plugin=qt
... lots of output ...
          Deploying dependencies for ELF file /home/user/Qt/5.12.3/gcc_64/lib/libicui18n.so.56 
DEBUG: Invalid ldd output:  	linux-vdso.so.1 (0x00007ffccddc0000) 
ERROR: Could not find dependency: libicuuc.so.56 
ERROR: Failed to deploy dependencies for existing files 
```

Indeed, when I check `ldd` by hand on just that dependency, what I get is:

```
$ ldd -d ~/Qt/5.12.3/gcc_64/lib/libicui18n.so.56 | grep found
	libicuuc.so.56 => not found
	libicudata.so.56 => not found
```

However, in my executable `libicuuc.so.56` and `libicudata.so.56` are definitely locatable:

```
$ ldd -d Example.AppDir/usr/bin/example | grep libicu
	libicui18n.so.56 => /home/user/Qt/5.12.3/gcc_64/lib/libicui18n.so.56 (0x00007fb31a2fc000)
	libicuuc.so.56 => /home/user/Qt/5.12.3/gcc_64/lib/libicuuc.so.56 (0x00007fb319f44000)
	libicudata.so.56 => /home/user/Qt/5.12.3/gcc_64/lib/libicudata.so.56 (0x00007fb318561000)
```

I'm not entirely sure how `ldd` actually resolves this library, but maybe somewhere down the line in Qt there might still be an RPATH being set or so?

---

TL;DR: linuxdeploy calls `ldd` directly on a dependency, which fails, while calling `ldd` on the executable that I'm trying to bundle resolves all dependencies recursively just fine.

**Comment by haampie** [comment]
Going further down the rabbit hole, running `readelf` on `libicui18n.so.56`, it seems it has some nonsensical RPATH set up:

```
$ readelf -d ~/Qt/5.12.3/gcc_64/lib/libicui18n.so.56 | grep path
 0x000000000000000f (RPATH)              Library rpath: [/home/qt/icu_install/lib]
```

this folder does not exist on my computer.

**Comment by TheAssassin** [comment]
Interesting bug. We actually run `ldd mybinary`, too, and collect dependencies listed there...

I wonder if `export LD_LIBRARY_PATH=.../AppDir/usr/lib` just fixed the issue, that would be an easy fix, as the binary is deployed first and should lead to libraries being found in that directory. That would, however, mean that we have to deploy the dependencies of binaries first, then run the search for additional libraries. That's however less efficient.

**Comment by haampie** [comment]
Hm yes, but wouldn't you agree it's completely sufficient to run `ldd -d mybinary`? `ldd` already recursively finds all shared libraries, and I guess `ldd` is the source of truth when it comes to resolving dependencies.

**Comment by TheAssassin** [comment]
I think that's what is being done. linuxdeploy is supposed to a) run on existing libraries and binaries (to support people using `make install` to install dependencies; libraries in `usr/lib/` must be deployed properly, too), b) deploys binaries passed via command line flags.

So, depends a bit. It's generally sufficient to run `ldd` on your binary to get the dependencies, but it's not generally sufficient to just do that. You can't just *not* deploy stuff in `usr/lib/`, there's that one use case where these libs are `dlopen()`d, therefore `ldd` won't list them and therefore their dependencies aren't caught unless you call `ldd` on them, too.

The question is, why does your main binary resolve the dependencies properly, but the rest does not? You need to set the `RUNPATH` on all your libraries, not just on the main binary. That might've worked for `RPATH` but certainly not for `RUNPATH`, as you say. You're the first person ever to show up with such a scenario, so my guess is the rest of the world have their `RUNPATH` just set on every library so that dependencies can be found.

Qt normally also have things set properly, we've been using both the Ubuntu trusty/xenial Qt distributions and Stephan Binner's Qt PPAs with great success.

**Comment by haampie** [comment]
Thanks for the detailed reply!

> So, depends a bit. It's generally sufficient to run ldd on your binary to get the dependencies, but it's not generally sufficient to just do that. You can't just not deploy stuff in usr/lib/, there's that one use case where these libs are dlopen()d, therefore ldd won't list them and therefore their dependencies aren't caught unless you call ldd on them, too.

But why is `ldd` being called on `libicu*` which is a direct dependency of libQtCore, which is in turn a direct dependency of my binary.

> The question is, why does your main binary resolve the dependencies properly, but the rest does not?

Yeah, this is really the question... I don't understand any of this. I did not find an RPATH set to the Qt libs dir anywhere, and the libicu* RPATHS are just bogus.

> Edit: oh, you're using your own Qt build. That indeed looks broken there.

I have actually just used the standard Qt Maintenance Tool to install Qt 5.12.3.

> Someone should carry this upstream to Qt...

I'll open an issue upstream indeed. Right now I've just changed the `RPATH`s in all icu libs to `RUNPATH` $ORIGIN, and linuxdeploy magically works again.

**Comment by TheAssassin** [comment]
I'm not saying we don't want to fix the issue, all I'm saying is that it's quite difficult for us due to different requirements, and I think technically it's the Qt/ICU stuff that's missing proper configuration.

> Yeah, this is really the question... I don't understand any of this. I did not find an RPATH set to the Qt libs dir anywhere, and the libicu* RPATHs are just bogus.

You just explained it. The Qt libraries don't have the rpath set. The Qt libraries are in a non-standard location.

The GNU linker used on most Linux distros takes into account several configuration values when trying to find the path of a library by its name. I don't know the exact order, but they are:

- locations trusted by the system (`/usr/lib/x86_64-linux-gnu`, /lib/x86_64-linux-gnu`...), these are configured in `/etc/ld.so.*`
- `$LD_LIBRARY_PATH` env var
- `RUNPATH` and sometimes even `RPATH`

Either of those must contain the path to the parent directory of a library. Your ICU stuff is not in a trusted system dir, it's not in the `RUNPATH`, so when running `ldd` on these dependencies, they can't be found.

Now, there's two ways to fix that. You demonstrated the first one, just fix the `RUNPATH` in the libraries and it will work. The other one would be to `export LD_LIBRARY_PATH=~/Qt/...` before running linuxdeploy. Since we use `ldd` internally, that'd also fix most issues.

I guess most people just do the latter, as linuxdeploy sets all `RUNPATH`s correctly, so when it's done everything is set up correctly.

> I have actually just used the standard Qt Maintenance Tool to install Qt 5.12.3.

Let's call that "non-standard" then. I guess most people just use their system Qt or the PPAs. We should probably add some docs on all these ways how to install Qt in the Qt plugin's tracker.

**Comment by haampie** [comment]
Hm, I don't immediately see how this explains that ldd finds ICU libs, since libicu is not a direct dependency of my own binary, so it must be a dependency of Qt's libs themselves. But none of the Qt libs specify RPATH [$ORIGIN], they only specify RUNPATH [$ORIGIN] (and RUNPATH is *not* inherited). So the chain would for instance be my_binary -> libQt5Core -> libicui18n -> libicuuc. In the last step from i18n -> uc, there is just a single, wrong RPATH, no RUNPATH, and no LD_LIBRARY_PATH.

Anways, I've opened a bug report upstream.

I'll try LD_LIBRARY_PATH next. One issue is that RPATH takes precedence over LD_LIBRARY_PATH however.

**Comment by TheAssassin** [comment]
What's the `RPATH`/`RUNPATH` of your main binary?

**Comment by haampie** [comment]
```
$ readelf -d example | grep path
 0x000000000000001d (RUNPATH)            Library runpath: [/usr/local/vtk8.1.1_gcc_64_release/lib:/usr/local/itk4.13.1_gcc_64_release/lib:/usr/local/minizip2.8.8_gcc_64_release/lib:/home/user/Qt/5.12.3/gcc_64/lib]
```

So, RUNPATH with a few dependencies that I'm using in different projects.

**Comment by TheAssassin** [comment]
Let us know if setting `LD_LIBRARY_PATH` fixes the issue.

**Comment by haampie** [comment]
Yes, it works as expected :)

Next Monday I might have some time to look into the source code of this package and see if I could improve anything. (The code quality looks amazing, nice job!) If anything, hopefully I can see why `ldd` is called on the dependency two levels deep -- if we can avoid calling `ldd` the whole bundling process might be a bit faster as well! (Already this package is much faster than linuxdeployqt for my application with many shared libraries as dependencies)

**Comment by TheAssassin** [comment]
As I said, `ldd` is called on all binaries inside `usr/lib` on every run. That applies to every consecutive run of linuxdeploy, and I think the binary deployment is run before we look for binaries in `usr/lib`, hence that also applies to the first run.

The idea of linuxdeploy is to be idempotent once it has run once successfully with a certain configuration. Also it is supposed to work with binaries installed via `make install`, see https://docs.appimage.org/packaging-guide/from-source/native-binaries.html#using-the-build-system-to-build-the-basic-appdir, it *must* run `ldd` again on all existing binaries in the AppDir to make sure it also deploys dependencies for libraries the application would `dlopen()`.

For performance reasons, linuxdeploy runs `ldd` on all binaries that shall be deployed, collects the information in some cache data structure (some sort of map) to avoid duplicate work and then finally does all the copying. You can observe that from the logs: all the copying and rpath magic occurs in the end.

The `ldd` bit is supposed to be replaced by something else, e.g., a custom binary dependency tracer. The algorithms are well defined in the end, and regular `ldd` can't be used for cross-compilation. There's an issue somewhere I think.

> The code quality looks amazing, nice job!

Thanks for the kind words, trying our best. The `AppDir` class IMO is a bit too loaded, but otherwise I'm quite happy with the quality. There is a need for more unit tests, though.

**Comment by haampie** [comment]
Ah, it starts to make sense now. I've only been using the package starting out with an empty `lib` folder, I agree things look differently when you start out with a `make install` structure.

Maybe this is not the right place to ask questions, but how do you support `dlopen` exactly? Isn't the issue there that the library name is only known at runtime?

**Comment by TheAssassin** [comment]
What I mean is, many people who use `dlopen` for their apps just install those libraries into the lib dir as well. If those libraries have dependencies, they need to be deployed, too. That's another scenario when you have to iterate over these libs, too: the binaries don't always directly depend on all dependencies.

There's some experimental tools to bundle `dlopen()`d libraries. It's quite simple sometimes, you just need to grep the source code for `*.so.*`, basically. But there's probably better ways, e.g., by analyzing the binary code.

**Comment by TheAssassin** [comment]
By the way, feel free to pass by in #appimage on Freenode.

**Comment by TheAssassin** [comment]
Closing as resolved.
